drivers/ipm/ipm_sedi.c

```diff
@@ -0,0 +1,298 @@
+{
+	struct ipm_sedi_context *ipm = dev->data;
+	unsigned int key = irq_lock();
+
+	atomic_set_bit(&ipm->status, is_write ? IPM_WRITE_BUSY_BIT : IPM_READ_BUSY_BIT);
+	pm_device_busy_set(dev);
+	irq_unlock(key);
+}
+
+static void clear_ipm_dev_busy(const struct device *dev, bool is_write)
+{
+	struct ipm_sedi_context *ipm = dev->data;
+	unsigned int key = irq_lock();
+
+	atomic_clear_bit(&ipm->status, is_write ? IPM_WRITE_BUSY_BIT : IPM_READ_BUSY_BIT);
+	if ((!atomic_test_bit(&ipm->status, IPM_WRITE_BUSY_BIT))
+		&& (!atomic_test_bit(&ipm->status, IPM_READ_BUSY_BIT))) {
+		pm_device_busy_clear(dev);
+	}
+	irq_unlock(key);
+}
+
+static void ipm_event_dispose(IN sedi_ipc_t device, IN uint32_t event, INOUT void *params)
+{
+	const struct device *dev = (const struct device *)params;
+	struct ipm_sedi_context *ipm = dev->data;
+	uint32_t drbl_in = 0, len;
+
+	LOG_DBG("dev: %u, event: %u", device, event);
+	switch (event) {
+	case SEDI_IPC_EVENT_MSG_IN:
+		if (ipm->rx_msg_notify_cb != NULL) {
+			set_ipm_dev_busy(dev, false);
+			sedi_ipc_read_dbl(device, &drbl_in);
+			len = IPC_HEADER_GET_LENGTH(drbl_in);
+			sedi_ipc_read_msg(device, ipm->incoming_data_buf, len);
+			ipm->rx_msg_notify_cb(dev,
+					      ipm->rx_msg_notify_cb_data,
+					      drbl_in, ipm->incoming_data_buf);
+		} else {
+			LOG_WRN("no handler for ipm new msg");
+		}
+		break;
+	case SEDI_IPC_EVENT_MSG_PEER_ACKED:
+		if (atomic_test_bit(&ipm->status, IPM_WRITE_IN_PROC_BIT)) {
+			k_sem_give(&ipm->device_write_msg_sem);
+		} else {
+			LOG_WRN("no sending in progress, got an ack");
+		}
+		break;
+	default:
+		return;
+	}
+}
+
+static int ipm_init(const struct device *dev)
+{
+	/* allocate resource and context*/
+	const struct ipm_sedi_config_t *info = dev->config;
+	sedi_ipc_t device = info->ipc_device;
+	struct ipm_sedi_context *ipm = dev->data;
+
+	info->irq_config();
+	k_sem_init(&ipm->device_write_msg_sem, 0, 1);
+	k_mutex_init(&ipm->device_write_lock);
+	ipm->status = 0;
+
+	sedi_ipc_init(device, ipm_event_dispose, (void *)dev);
+	atomic_set_bit(&ipm->status, IPM_PEER_READY_BIT);
+	LOG_DBG("ipm driver initialized on device: %p", dev);
+	return 0;
+}
+
+static int ipm_send_isr(const struct device *dev,
+			 uint32_t drbl,
+			 const void *msg,
+			 int msg_size)
+{
+	const struct ipm_sedi_config_t *info = dev->config;
+	sedi_ipc_t device = info->ipc_device;
+	uint32_t drbl_acked = 0;
+
+	sedi_ipc_write_msg(device, (uint8_t *)msg,
+			   (uint32_t)msg_size);
+	sedi_ipc_write_dbl(device, drbl);
+	do {
+		sedi_ipc_read_ack_drbl(device, &drbl_acked);
+	} while ((drbl_acked & BIT(IPC_BUSY_BIT)) == 0);
+
+	return 0;
+}
+
+static int ipm_sedi_send(const struct device *dev,
+		     int wait,
+		     uint32_t drbl,
+		     const void *msg,
+		     int msg_size)
+{
+	__ASSERT((dev != NULL), "bad params\n");
+	const struct ipm_sedi_config_t *info = dev->config;
+	struct ipm_sedi_context *ipm = dev->data;
+	sedi_ipc_t device = info->ipc_device;
+	int ret = -1;
```

Comment:
Please don't mix `-1` with POSIX error values (like `-EINVAL` or `-ENOTSUP`). I think it would be better to not initialize the variable upon declaration, and instead set it explicitly to a meaningful value in each error branch (you're already doing this in most branches, but not all).

Reply:
switched sedi error to posix ones